### DISCUSSION TITLE
Connection DPY-4011 exception

### DISCUSSION THREAD

**Discussion by Vethorm** [oracle/python-oracledb#45] [Q&A]
Hello,

I'm currently working with Apache Airflow and the newer version we are upgrading to is using oracledb over cx_oracle. Previously we had no issue connecting with cx_oracle and still don't. Oracledb however seems to hang and then throw DPY-4011 exceptions. I have't been able to find much information on this aside from 1 or 2 articles mentioning to use `disable_oob=True` which unfortunately I can't set with Apache Airflow in the current version. I have confirmed this does fix our issue however.

I'm wondering if this is something I can resolve without the connection parameter `disable_oob=True`.

Happy to provide anymore information if it helps.

Thanks!

**Comment by anthony-tuininga** [comment]
You can in your application (before it creates a connection) make a call to `init_oracle_client()` -- this will enable thick mode and behaves the same way as cx_Oracle did.

We would, however, like to resolve the issue for thin mode, too. Can you provide details on the database version and the nature of the network connection between them?

**Reply by Vethorm** [reply]
Doing a `SELECT * FROM V$VERSION` on SQL developer for the database I'm trying to connect to shows
`Database 19c Standard Edition 2 Release 19.0.0.0.0 - ProductionVersion 19.14.0.0.0`

The oracledb version is 1.0.1

As far as the network connection between them. This is on a company network so I'm unsure of the specifics. Is there a more specific question about it I can ask my DBA?

**Reply by anthony-tuininga** [reply]
Can you verify that the problem still exists with oracledb 1.0.3? Is the client on the same network as the database? Can you also run your script after setting the environment variable `PYO_DEBUG_PACKETS` to any value and post the results here? It is possible that the database is sending notice that OOB should not be enabled and the thin driver isn't catching that -- the packet data may help with that. What platforms are you using for both client and database?

**Reply by Vethorm** [reply]
> oracledb 1.0.3

I've upgraded the package and still receive the same issue

> client on the same network as database

I believe it should be, but I can double check with my DBA tomorrow

> Platforms for both client and database

The machine running the script is CentOS 7 and also a docker container on Debian 11
The database I believe is running on AWS RDS

> `PYO_DEBUG_PACKETS`

I've removed some connection information from the packets

```2022-08-09 19:36:29.352 [socket: 3] Sending packet:
0000 : 01 17 00 00 01 00 00 00 |........|
0008 : 01 3E 01 2C 0C 01 20 00 |.>.,....|
0016 : FF FF 4F 98 00 00 00 01 |..O.....|
0024 : 00 CD 00 4A 00 00 00 00 |...J....|
0032 : 80 80 00 00 00 00 00 00 |........|
0040 : 00 00 00 00 00 00 00 00 |........|
0048 : 00 00 00 00 00 00 00 00 |........|
0056 : 00 00 00 00 20 00 00 00 |........|
0064 : FF FF 00 00 00 00 00 00 |........|
0072 : 00 01 28 44 45 53 43 52 |..(DESCR|
0080 : 49 50 54 49 4F 4E 3D 28 |IPTION=(|
0088 : 41 44 44 52 45 53 53 3D |ADDRESS=|
0096 : 28 50 52 4F 54 4F 43 4F |(PROTOCO|
0104 : 4C 3D 54 43 50 29 28 48 |L=TCP)(H|
0112 : 4F 53 54 3D             |OST=    |
0120 :          29 28 50 4F 52 |   )(POR|
0128 : 54 3D             29 29 |T=    ))|
0136 : 28 43 4F 4E 4E 45 43 54 |(CONNECT|
0144 : 5F 44 41 54 41 3D 28 53 |_DATA=(S|
0152 : 45 52 56 49 43 45 5F 4E |ERVICE_N|
0160 : 41 4D 45 3D             |AME=    |
0168 :       29 28 43 49 44 3D |  )(CID=|
0176 : 28                      |(       |
0184 :                         |        |
0192 :                         |        |
0200 :                         |        |
0208 :                         |        |
0216 :                         |        |
0224 :       29 28             |  )(    |
0232 : 3D                      |=       |
0240 :                         |        |
0248 :                         |        |
0256 :                   29 28 |      )(|
0264 : 55 53 45 52 3D          |USER=   |
0272 :          29 29 29 XX    |   )))) |

2022-08-09 19:36:29.390 [socket: 3] Receiving packet:
0000 : 00 08 00 00 0B 00 00 00 |........|

2022-08-09 19:36:29.390 [socket: 3] Sending packet:
0000 : 01 17 00 00 01 00 00 00 |........|
0008 : 01 3E 01 2C 0C 01 20 00 |.>.,....|
0016 : FF FF 4F 98 00 00 00 01 |..O.....|
0024 : 00 CD 00 4A 00 00 00 00 |...J....|
0032 : 80 80 00 00 00 00 00 00 |........|
0040 : 00 00 00 00 00 00 00 00 |........|
0048 : 00 00 00 00 00 00 00 00 |........|
0056 : 00 00 00 00 20 00 00 00 |........|
0064 : FF FF 00 00 00 00 00 00 |........|
0072 : 00 01 28 44 45 53 43 52 |..(DESCR|
0080 : 49 50 54 49 4F 4E 3D 28 |IPTION=(|
0088 : 41 44 44 52 45 53 53 3D |ADDRESS=|
0096 : 28 50 52 4F 54 4F 43 4F |(PROTOCO|
0104 : 4C 3D 54 43 50 29 28 48 |L=TCP)(H|
0112 : 4F 53 54 3D             |OST=    |
0120 :          29 28 50 4F 52 |   )(POR|
0128 : 54 3D             29 29 |T=    ))|
0136 : 28 43 4F 4E 4E 45 43 54 |(CONNECT|
0144 : 5F 44 41 54 41 3D 28 53 |_DATA=(S|
0152 : 45 52 56 49 43 45 5F 4E |ERVICE_N|
0160 : 41 4D 45 3D             |AME=    |
0168 :       29 28 43 49 44 3D |  )(CID=|
0176 : 28                      |(       |
0184 :                         |        |
0192 :                         |        |
0200 :                         |        |
0208 :                         |        |
0216 :                         |        |
0224 :       29 28             |  )(    |
0232 : 3D                      |=       |
0240 :                         |        |
0248 :                         |        |
0256 :                   29 28 |      )(|
0264 : 55 53 45 52 3D          |USER=   |
0272 :          29 29 29 XX    |   )))) |

2022-08-09 19:36:29.419 [socket: 3] Receiving packet:
0000 : 00 2D 00 00 02 00 00 00 |.-......|
0008 : 01 3E 0C 01 00 00 00 00 |.>......|
0016 : 01 00 00 00 00 2D C1 00 |.....-..|
0024 : 00 00 00 00 00 00 00 00 |........|
0032 : 00 00 20 00 00 00 FF FF |........|
0040 : 00 00 00 00 01          |.....   |

2022-08-09 19:36:29.419 [socket: 3] Sending packet:
0000 : 00 00 00 0B 0C 00 00 00 |........|
0008 : 01 00 02                |...     |

2022-08-09 19:36:29.419 [socket: 3] Sending packet:
0000 : 00 00 00 8F 06 00 00 00 |........|
0008 : 00 00 DE AD BE EF 00 85 |........|
0016 : 0B 20 02 00 00 04 00 00 |........|
0024 : 04 00 03 00 00 00 00 00 |........|
0032 : 04 00 05 0B 20 02 00 00 |........|
0040 : 08 00 01 00 00 10 1C 66 |.......f|
0048 : EC 28 EA 00 12 00 01 DE |.(......|
0056 : AD BE EF 00 03 00 00 00 |........|
0064 : 04 00 04 00 01 00 02 00 |........|
0072 : 03 00 01 00 03 00 00 00 |........|
0080 : 00 00 04 00 05 0B 20 02 |........|
0088 : 00 00 02 00 03 E0 E1 00 |........|
0096 : 02 00 06 FC FF 00 02 00 |........|
0104 : 02 00 00 00 00 00 04 00 |........|
0112 : 05 0B 20 02 00 00 01 00 |........|
0120 : 01 00 00 03 00 02 00 00 |........|
0128 : 00 00 00 04 00 05 0B 20 |........|
0136 : 02 00 00 01 00 01 00    |....... |```

**Reply by anthony-tuininga** [reply]
Interesting! I see from the packet trace that OOB is supported by the server -- but it is still dropping the connection! Are you able to get server traces? That might help us figure out what is going on here. I am unable to replicate this myself.

One other option: if OOB isn't actually supported by your network configuration, you can also add this line in a *server* side sqlnet.ora:

```
disable_oob=on
```

That will turn it off at the server level and there won't be any attempt to negotiate such support.

**Reply by anthony-tuininga** [reply]
Another possible way to figure this out:
- set up *client* side tracing in your client side sqlnet.ora
- call `init_oracle_client()` to enable thick mode

That information may also help me figure out what the thick client is doing in your situation (since I am unable to replicate it at the moment).

**Reply by Vethorm** [reply]
Thanks for the help so far

Unfortunately we aren't able to use `DISABLE_OOB=ON` since we are using managed Amazon RDS and AWS doesn't support that option in their console. Our DBA had opened a ticket in the past and they basically told us it wasn't supported it.

I'm also currently unable to set `disable_oob=True` with Airflow since the Airflow hook doesn't support that option currently and is quite limited. This might require raising an issue with the provider package.
https://github.com/apache/airflow/blob/main/airflow/providers/oracle/hooks/oracle.py#L60

I'll try and get client side tracing setup today or tomorrow and post results. Would client side tracing with cx_Oracle work or does it need to be with oracledb.

**Reply by anthony-tuininga** [reply]
Tracing with cx_Oracle would be acceptable as well. python-oracledb thick mode is the same as cx_Oracle in most respects.

**Reply by Vethorm** [reply]
After looking on how to setup client side tracing I'm somewhat unsure how to do so. Are there any resources or instructions on how to do so?

**Reply by anthony-tuininga** [reply]
Another option: a bit of a hack but may work for you?

```python
import oracledb

orig_connect = oracledb.connect

def my_connect(*args, **kwargs):
    print("calling my connection!")
    return orig_connect(*args, disable_oob=True, **kwargs)

oracledb.connect = my_connect
```

You would need to do this before Apache Airflow is invoked -- not sure if that is possible?

**Reply by Vethorm** [reply]
This I hadn't thought of haha. It actually seems to work for Airflow though when defined in my DAG. This will get things at least working for me at least.

I will most likely raise an issue with the Airflow provider so the option can be given with the hack, but will use this in the meantime.

Do you believe there might be an issue with the thin client or is this just a matter of needing to disable oob.

**Reply by anthony-tuininga** [reply]
It is entirely possible that there is an issue with the thin client. Or it is possible that the thick client simply detects the error and retries with `disable_oob = on`. The client trace should help to figure out exactly what the thick driver is doing. I've also asked internally and will get back to you if I get an answer on that front.

**Reply by anthony-tuininga** [reply]
What version of the Oracle Client libraries are you using for the thick driver? You can do this to find out:

```python
oracledb.init_oracle_client()
print(oracledb.clientversion())
```

**Reply by anthony-tuininga** [reply]
Also, can you provide details on the host you are using? Is it "localhost" or an actual name? Judging from the amount of spaces I think I can answer that question -- but thought I would make sure!

**Reply by Vethorm** [reply]
Oracle client library version: 12.2.0.1.0
The host is an actual name to an Oracle AWS RDS instance

**Reply by anthony-tuininga** [reply]
Ah! That does not have the OOB check included in it. Can you try with a 19c or 21c client which does have that check? You can use an instant client zip file and set ``LD_LIBRARY_PATH`` to the unzipped location.
